### PR TITLE
Replace timer based RNG seed with ADC reference based one

### LINKED ISSUE #115: Init RNG seeding seemingly very predictable on Arduino Uno
I wrote a test program which does hydro_init(), and then writes the random state over serial.
In the first test, I did soft resets with the RESET pin, and in the second test I disconnected power manually for a few seconds to reset. As you can see, most repeats are close to each other, but in some cases the same state appears again several minutes of testing later. In the soft reset case, there are only 5 different states in total, the least common of which appears 3 times, and the most common one appears 35 times. In the hard reset case, there are 18 different states, of which 10 appear only once each, and the most common state appears 9 times.

https://paste.sr.ht/~lonjil/a9f50c3e28beb4cf26f1632f3caef315e859974b

Am I perhaps doing something wrong? Is there something I could do to mitigate this?

### PR DESCRIPTION
**Pull request by DeeFuse**
Remove "Arduino.h" dependency
Fixes #115

### CODE REVIEW

**Comment by lonjil** [comment]
I changed it so that it sets it to low right at the end of the function, and sets it to high right before sampling. This seems to increase the number of unique readings.

**Comment by lonjil** [comment]
I generated 1.6GiB. It has 100 different distinct byte values. I'm currently writing a tool in Zig to extract and analyze it.

**Comment by lonjil** [comment]
I ran a Von Neumann extractor on just the lowest bit of each byte, and the output was pretty even according to `ent`:
```
Entropy = 7.984414 bits per byte.

Optimum compression would reduce the size
of this 31211520 byte file by 0 percent.

Chi square distribution for 31211520 samples is 673722.13, and randomly
would exceed this value less than 0.01 percent of the times.

Arithmetic mean value of data bytes is 127.5149 (127.5 = random).
Monte Carlo value for Pi is 3.187330063 (error 1.46 percent).
Serial correlation coefficient is 0.001175 (totally uncorrelated = 0.0).
```
It also passes a few dieharder tests, but the size (30M) is not big enough for most of the tests.
Doing an extraction on every bit of every byte gave quite a crappy result. Next I'm going to separately test each other bit position in the same manner.


I'm currently running another arduino sketch that generates 4096 samples and then resets, to see if there are any right-after-boot corelations between runs.

**Comment by lonjil** [comment]
I have now done Von Neumann extraction on each bit position.
Here are the sizes of each output file:
```
Bits     MB   File             Output size / Input size
31211520  30M von_neumann_bit0 0.14922
23101440  23M von_neumann_bit1 0.11045
18657280  18M von_neumann_bit2 0.08920
18440192  18M von_neumann_bit3 0.08816
 1937408 1.9M von_neumann_bit4 0.00926
  962560 940K von_neumann_bit5 0.00460
  372736 364K von_neumann_bit6 0.00178
  323584 316K von_neumann_bit7 0.00155
```
According to ent [1] each of these files is high entropy. Combined with the decrease in file sizes, I interpret this as meaning that while the high bits are very biased, they are still quite decorrelated. This doesn't mean that they're independent, but I think it's a good sign. I think an iterated Von Neumann extractor might be useful to get more examinable data out of the raw data, but I'm not sure.


[1] https://paste.sr.ht/~lonjil/b6814c2139c2f55c9d37abece27aab44bc568823

**Comment by jedisct1** [comment]
This is great!

How long does it take to extract 256 bits from the lower bit?

**Comment by lonjil** [comment]
On my phone so my numbers are inexact, but I believe it took around 1707
input bits for Von Neumann to produce 256 bits.

**Comment by lonjil** [comment]
As a point of reference regarding performance, this hydro_random_init [1] implementation takes ~517 milliseconds to run `hydro_init()`. 2048 samples, which I believe means 131 (?) or so invocations of the Gimli function. With 1024 samples, it takes ~268 milliseconds. 4096 samples ends up at ~1014 milliseconds.

[1] https://paste.sr.ht/~lonjil/4a2cd787141af107fb5ac3b7eb3914b9d2034eef

**Comment by DeeFuse** [comment]
I see that you removed the ADC [converisontime "randomisation"](https://github.com/jedisct1/libhydrogen/blob/878e59d1ca24d12ce07e695baa0fc831fbdb925f/impl/random/avr.h#L44).
With that in place the randomnes of the readings could be increased.

**Comment by lonjil** [comment]
I'm not sure it would. Changing the scaler to 2 makes it not random at all,
and changing it to 128 makes it much slower while also reducing the entropy
per reading (at a glance anyway, I didn't do a rigorous test). I think the
other possible selections other than 64 all produced less entropy as well,
so any extra randomness from an unpredictable ADC clock might just be lost
from a more predictable sample output in some of the cases.

Also, the clock scaler needs to be set to a different value depending on
the chip clock rate. An atmega328p at 8mhz should probably use 32, one at
4mhz should probably use 16, and so on, and one at 20mhz might do best at
128. The random selection would make that aspect harder to reason about,
and not possible at all on some configurations.

**Comment by lonjil** [comment]
I ran an implemention of NIST 800-90B non-IID min-entropy estimator on 2 million samples, and it gave 0.315568 bits per sample.

**Comment by DeeFuse** [comment]
I've tested the codes with a ATmega2561 in a comercial appliance I had at hand.
Following porblems emerged:

1. The V_BG ADMUX settings are different (compared to the ATmega328)
2. External (strong) AREF pullups will result in identical readings every call (with @lonjil code as well as wiht the StackOverflow suggestion)

**Comment by lonjil** [comment]
Since different AVR micros differ so strongly on so many points, it may be necessary to have bespoke routines for every model :/

**Comment by jedisct1** [comment]
This is sad :(

And we won't be able to test on every AVR model.

**Comment by lonjil** [comment]
I guess the code will have to have preprocessor directives for the target chip, and an error instructing people to open an issue for un-implemented models.

Also, I wonder if there is a published in-depth description of the ADC hardware anywhere, so that someone with the appropriate physics knowledge could figure out a lower bound on the min-entropy, rather than the upper bound we get from sample testing.

**Comment by DeeFuse** [comment]
I think we won't be able to determine a definite min-entropy bound for that type of source.
After running the tests on my boards I think this aproach is not fitting for this librarys goar of beeing ["...hard-to-misuse..."](https://github.com/jedisct1/libhydrogen/blame/master/README.md#L9) when it is possible to accidentially "disable" the random data source by applying a "common hardware design"
TL;DR: I think we should not rely on external circumstances / hardware designs for a source of random data

I'm currently testing an other (WDT + Timer based) RNG which woudl be independent of external / user influencable circumstances. The randomnes here will come from the "low accuracy" of the internal 128kHz RC oscillator for the WDT.

```
Entropy = 7.957624 bits per byte.

Optimum compression would reduce the size
of this 4096 byte file by 0 percent.

Chi square distribution for 4096 samples is 239.62, and randomly
would exceed this value 74.71 percent of the times.

Arithmetic mean value of data bytes is 127.8352 (127.5 = random).
Monte Carlo value for Pi is 3.131964809 (error 0.31 percent).
Serial correlation coefficient is 0.031658 (totally uncorrelated = 0.0).
```

**Comment by lonjil** [comment]
@DeeFuse would you mind posting the code for that?

**Comment by DeeFuse** [comment]
@lonjil You can find the code I've used [here](https://gist.github.com/DeeFuse/9eccade5aa0b8fb2557a9ea17916403c)  (cleaned up)
My test hardware was a ATmega2561 with F_CPU = 14745600UL

**Comment by jedisct1** [comment]
Old PR and discussion, but still relevant :)

What should we do?

**Comment by DeeFuse** [comment]
I could clean up my GIST with the updated code and push it to my PR.
I'm no security / crypto expert. How do you feel about the implementation of randomnet used in my code?

**Comment by jedisct1** [comment]
The implementation looks fine, but I don't have the hardware to test it on.

**Comment by DeeFuse** [comment]
If you want to have a look into the randomnes of this implementation I can send you a bin file for analysis.

**Comment by justind000** [comment]
> The implementation looks fine, but I don't have the hardware to test it on.

I have a lot of different boards available and can help out if needed.

**Comment by DeeFuse** [comment]
I created a 10k number sequence as well as 10 shorter dumps after booting the MCU: [random.zip](https://github.com/jedisct1/libhydrogen/files/8429338/random.zip)
I've done some reading about the AVR hardware differences which need to be accounted for to make it foolproof to use it.
After implementing some #ifdef to catch those hardware differences I'll feel comfortable to push / merge the code into libhydrogen.
@justind000 I've updated my [gist](https://gist.github.com/DeeFuse/9eccade5aa0b8fb2557a9ea17916403c) with the latest code. Feel free to run it on your AVRs. The random numbers are printed to USART1 @115200 baud

**Comment by DeeFuse** [comment]
Shall we backup the states of the used perihpherals (WDT / TIM1) and restore them after the random number generation?

**Comment by jedisct1** [comment]
> Shall we backup the states of the used perihpherals (WDT / TIM1) and restore them after the random number generation?

That sounds like a good idea. Applications may not expect them to be changed.

**Comment by DeeFuse** [comment]
I got a question about the coding style:
I can calculate the required timer parameters at compile time or at runtime.
The main difference is that the code when compilet with -01 or higher will be equal but at -O0 there is a huge difference.
The drawback of the precompiler calculation is the readability.
Which solution shall I include? test1 - Preprocessor / test2 - runtime (or optimized) https://godbolt.org/z/z37E6ox6s

**Comment by DeeFuse** [comment]
I'm thinking about disabling the function 'hydro_random_reseed' when building on an AVR unless the user explicitly enables it.
If the user calls this function while running their own code it could cause a lot of problems due to the reconfiguration of TIM1 and the ~4second execution time.
